nearcache_map.go

```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		ncm.nc.Invalidate(key)
+		return nil, err
+	}
+	return value, nil
+}
+
+func (ncm *nearCacheMap) Put(ctx context.Context, m *Map, key, value interface{}, ttl int64) (interface{}, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return nil, err
+	}
+	defer ncm.nc.Invalidate(key)
+	return m.putWithTTLFromRemote(ctx, key, value, ttl)
+}
+
+func (ncm *nearCacheMap) PutWithMaxIdle(ctx context.Context, m *Map, key, value interface{}, ttl int64, maxIdle int64) (interface{}, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return nil, err
+	}
+	defer ncm.nc.Invalidate(key)
+	return m.putWithMaxIdleFromRemote(ctx, key, value, ttl, maxIdle)
+}
+
+func (ncm *nearCacheMap) Remove(ctx context.Context, m *Map, key interface{}) (interface{}, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return false, err
+	}
+	defer ncm.nc.Invalidate(key)
+	return m.removeFromRemote(ctx, key)
+}
+
+func (ncm *nearCacheMap) RemoveIfSame(ctx context.Context, m *Map, key interface{}, value interface{}) (bool, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return false, err
+	}
+	defer ncm.nc.Invalidate(key)
+	return m.removeIfSameFromRemote(ctx, key, value)
+}
+
+func (ncm *nearCacheMap) Set(ctx context.Context, m *Map, key, value interface{}, ttl int64) error {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return err
+	}
+	defer ncm.nc.Invalidate(key)
+	return m.setFromRemote(ctx, key, value, ttl)
+}
+
+func (ncm *nearCacheMap) SetWithTTLAndMaxIdle(ctx context.Context, m *Map, key, value interface{}, ttl time.Duration, maxIdle time.Duration) error {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return err
+	}
+	defer ncm.nc.Invalidate(key)
+	return m.setWithTTLAndMaxIdleFromRemote(ctx, key, value, ttl, maxIdle)
+}
+
+func (ncm *nearCacheMap) TryRemove(ctx context.Context, m *Map, key interface{}, timeout int64) (interface{}, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return false, err
+	}
+	defer ncm.nc.Invalidate(key)
+	return m.tryRemoveFromRemote(ctx, key, timeout)
+}
+
+func (ncm *nearCacheMap) GetLocalMapStats() LocalMapStats {
+	return LocalMapStats{
+		NearCacheStats: ncm.nc.Stats(),
+	}
+}
+
+func (ncm *nearCacheMap) getCachedValue(key interface{}, deserialize bool) (value interface{}, found bool, err error) {
+	value, found, err = ncm.nc.Get(key)
+	if err != nil {
+		return nil, false, err
+	}
+	if !found {
+		return nil, false, nil
+	}
+	if value == nil {
+		return nil, true, nil
+	}
+	if deserialize {
+		data, ok := value.(serialization.Data)
+		if ok {
+			value, err = ncm.ss.ToObject(data)
+			if err != nil {
+				return nil, false, err
+			}
+		}
+	}
+	return value, true, nil
+}
+
+func (ncm *nearCacheMap) getFromRemote(ctx context.Context, m *Map, key interface{}) (interface{}, error) {
+	keyData, err := m.validateAndSerialize(key)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
You can move this serialization code to `m.getFromRemote` as well. I missed this before.

Reply:
`keyData` is passed to `TryReserveForUpdate` at line 215. So `key` would be serialized twice if we move that code to `getFromRemote`.

Reply:
ok sorry.